### ISSUE TITLE
Use PyTorch and TensorFlow to test CAB

### ISSUE THREAD

**Issue by mm6234** [xai-privacy/compliance-audit-benchmark#2]
**Tool Definitions:**
- PyTorch and TensorFlow are open-source libraries for building and training machine learning models, especially neural networks, using efficient numerical computations with tensors.
*(These tools can also be used for creating a baseline model with custom training data for benchmark testing.)
- Differential privacy - mathematical guarantee that a machine learning model does not reveal whether any single individual’s data was used in training.

**How can PyTorch/TensorFlow be used for measuring the effectiveness of the benchmark?**
1. Create privacy-aware baselines
a. Models can be trained with DP-SGD (TensorFlow Privacy) or Opacus (PyTorch) to make them "differentially private" (https://arxiv.org/pdf/1607.00133, pages 1 & 5).
b. These models can be compared  to regular models to measure the trade-off between privacy and compliance/leakage metrics (https://arxiv.org/pdf/1607.00133, page 9.)
2. Provide privacy accounting
a. Both libraries track formal privacy metrics like ε (epsilon) and δ (delta) for trained models, providing evidence that training follows regulations (https://www.tensorflow.org/responsible_ai/privacy/tutorials/classification_privacy?utm_source=chatgpt.com).
3. Generate synthetic data
a. These tools can create mock health or financial data through synthetic datasets that mimic real data while protecting privacy (TensorFlow Privacy library).
b. IBM's diffprivlib Python library can also be used to create synthetic data (https://arxiv.org/pdf/1907.02444, https://diffprivlib.readthedocs.io/en/latest/?utm_source=chatgpt.com)
4. Measure mitigation effectiveness
a. Run attacks against differentially-protected (DP) and non-DP models to quantify how much DP reduces attack successes (https://arxiv.org/pdf/1610.05820, https://www.usenix.org/system/files/sec21-carlini-extracting.pdf?utm_source=chatgpt.com, https://www.usenix.org/system/files/sec21fall-song.pdf).
    * Standardized privacy attacks (membership inference or model extraction).

**Tool Limitations**
- There is a trade-off between privacy and performance. DP can reduce model accuracy.
- Synthetic data may not capture real-world complexity.
- Privacy accounting works best on training models, but doesn't prevent all possible leaks in deployed LLM systems.

**Comment by SebastianZimmeck** [comment]
Thanks, @mm6234. A few initial thoughts:

- Testing via PyTorch and TensorFlow would only work for our own models, right? Otherwise, how would we integrate those libraries into, say, Gemini?
- In my mind, the testing is mostly manual, i.e., experts/we provide the ground truth against which we compare the model output (and we measure agreement among experts to ensure the ground truth is reliable, e.g., using Krippendorff's Alpha).
- We can use Differential Privacy, but that is a big discussion to have.
